python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py

```diff
@@ -0,0 +1,370 @@
+    we set the timestamp to 0.0 so that the timestamps of other parent materializations will take precedent.
+
+    Here's how we get there:
+
+    We want to get the materialization information for all of the parents of an asset to determine
+    if we want to materialize the asset in this sensor tick. We also need determine the new cursor
+    value for the asset so that we don't process the same materialization events for the parent
+    assets again.
+
+    We iterate through each parent of the asset and determine its materialization info. The parent
+    asset's materialization status can be one of three options:
+    1. The parent has materialized since the last time the child was materialized (determined by comparing
+        the timestamp of the parent materialization to the cursor_timestamp).
+    2. The parent will be materialized as part of the materialization that will be kicked off by the
+        sensor.
+    3. The parent has not been materialized and will not be materialized by the sensor.
+
+    In cases 1 and 2 we indicate that the parent has been updated by setting its value in
+    parent_asset_event_records to True. For case 3 we set its value to False.
+
+    If wait_for_in_progress_runs=True, there is another condition we want to check for.
+    If any of the parents is currently being materialized we want to wait to materialize current_asset
+    until the parent materialization is complete so that the asset can have the most up to date data.
+    So, for each parent asset we check if it has a planned asset materialization event in a run that
+    is currently in progress. If this is the case, we don't want current_asset to materialize, so we
+    set parent_asset_event_records to False for all parents (so that if the sensor is set to
+    materialize if any of the parents are updated, the sensor will still choose to not materialize
+    the asset) and immediately return.
+    """
+    from dagster._core.events import DagsterEventType
+    from dagster._core.storage.event_log.base import EventRecordsFilter
+
+    parent_asset_event_records: Dict[AssetKey, Tuple[bool, float]] = {}
+
+    for p in parent_assets:
+        if p in will_materialize_list:
+            # if p will be materialized by this sensor, then we can also materialize current_asset
+            # we don't know what time asset p will be materialized so we set the cursor val to 0.0
+            parent_asset_event_records[p] = (
+                True,
+                0.0,
+            )
+        # TODO - when source asset versioning lands, add a check here that will see if the version has
+        # updated if p is a source asset
+        else:
+            if wait_for_in_progress_runs:
+                # if p is currently being materialized, then we don't want to materialize current_asset
+
+                # get the most recent planned materialization
+                materialization_planned_event_records = context.instance.get_event_records(
```

Comment:
Will this result in sometimes querying for the same asset key multiple times. If so, could we cache instead?

Reply:
like if asset A and B both have a parent C? in that case yeah when we get the parent info for A we'll fetch the event for C and then when we get the parent info for B we'll make the same call for C. Definitely seems like a good place for caching, i'll add that

Reply:
right